dfc/metasync_test.go

```diff
@@ -669,3 +669,257 @@ func TestMetaSyncMembership(t *testing.T) {
+	chTarget := make(chan map[string]string, 10)
+	fTarget := func(w http.ResponseWriter, r *http.Request) {
+		d := make(map[string]string)
+		primary.readJSON(w, r, &d) // ignore json error
+		chTarget <- d
+	}
+
+	s := httptest.NewServer(http.HandlerFunc(fProxy))
+	defer s.Close()
+	primary.smap.addProxy(&proxyInfo{daemonInfo{DaemonID: "proxy", DirectURL: s.URL}, false /* primary */})
+	proxy := proxyrunner{}
+	proxy.smap = &Smap{Tmap: make(map[string]*daemonInfo), Pmap: make(map[string]*proxyInfo)}
+	proxy.lbmap = newLBMap()
+
+	// empty payload
+	newSMap, ooldSMap, actMsg, errStr := proxy.extractsmap(make(map[string]string))
+	if newSMap != nil || ooldSMap != nil || actMsg != nil || errStr != "" {
+		t.Fatal("Extract smap from empty payload returned data")
+	}
+
+	syncer.sync(true, primary.smap.cloneU())
+	payload := <-chProxy
+
+	newSMap, ooldSMap, actMsg, errStr = proxy.extractsmap(payload)
+	noErr(errStr)
+	emptyActionMsg(actMsg)
+	emptySMap(ooldSMap)
+	matchSMap(primary.smap, newSMap)
+	proxy.smap = newSMap
+
+	// same version of smap received
+	newSMap, ooldSMap, actMsg, errStr = proxy.extractsmap(payload)
+	noErr(errStr)
+	emptyActionMsg(actMsg)
+	emptySMap(ooldSMap)
+	nilSMap(newSMap)
+
+	// older version of smap received
+	proxy.smap.Version++
+	_, _, _, errStr = proxy.extractsmap(payload)
+	hasErr(errStr)
+	proxy.smap.Version--
+
+	var am ActionMsg
+	y := &ActionMsg{"", "", primary.smap}
+	b, _ := json.Marshal(y)
+	json.Unmarshal(b, &am)
+	prevSMap := primary.smap.cloneU()
+
+	s = httptest.NewServer(http.HandlerFunc(fTarget))
+	defer s.Close()
+	primary.smap.add(&daemonInfo{DaemonID: "target", DirectURL: s.URL})
+	target1 := proxyrunner{}
+	target1.smap = &Smap{Tmap: make(map[string]*daemonInfo), Pmap: make(map[string]*proxyInfo)}
+	target1.lbmap = newLBMap()
+
+	c := func(n, o *Smap, m *ActionMsg, e string) {
+		noErr(e)
+		matchActionMsg(&am, m)
+		matchPrevSMap(prevSMap, o)
+		matchSMap(primary.smap, n)
+	}
+
+	syncer.sync(true, primary.smap.cloneU())
+	payload = <-chProxy
+	newSMap, ooldSMap, actMsg, errStr = proxy.extractsmap(payload)
+	c(newSMap, ooldSMap, actMsg, errStr)
+
+	payload = <-chTarget
+	newSMap, ooldSMap, actMsg, errStr = target1.extractsmap(payload)
+	c(newSMap, ooldSMap, actMsg, errStr)
+
+	// note: this extra call is not necessary by meta syncer
+	// it picked up the target as it is new added target, but there is no other meta needs to be synced
+	// (smap is the only one and it is already synced)
+	payload = <-chTarget
+
+	y = &ActionMsg{"", "", primary.smap}
+	b, _ = json.Marshal(y)
+	json.Unmarshal(b, &am)
+	prevSMap = primary.smap.cloneU()
+
+	s = httptest.NewServer(http.HandlerFunc(fTarget))
+	defer s.Close()
+	primary.smap.add(&daemonInfo{DaemonID: "target2", DirectURL: s.URL})
+	target2 := proxyrunner{}
+	target2.smap = &Smap{Tmap: make(map[string]*daemonInfo), Pmap: make(map[string]*proxyInfo)}
+	target1.lbmap = newLBMap()
+
+	c = func(n, o *Smap, m *ActionMsg, e string) {
```

Comment:
What is the difference between this `c` and `c` at line 825?

Reply:
haha, nice catch, use to be different, now they are the same, thank you